test/e2e/pkg/grammar/checker.go

```diff
@@ -0,0 +1,215 @@
+	}
+	return parts[0]
+}
+
+// getExecutionString returns a string of terminal symbols in parser readable format.
+func (g *GrammarChecker) getExecutionString(reqs []*abci.Request) string {
+	s := ""
+	for _, r := range reqs {
+		t := g.getRequestTerminal(r)
+		// We ensure to have one height per line for readability.
+		if t == COMMIT {
+			s += t + "\n"
+		} else {
+			s += t + " "
+		}
+	}
+	return s
+}
+
+// Verify verifies whether a list of request satisfy ABCI grammar.
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	if len(reqs) == 0 {
+		return false, fmt.Errorf("execution with no ABCI calls.")
+	}
+	r := g.filterRequests(reqs)
+	// Check if the execution is incomplete.
+	if len(r) == 0 {
+		return true, nil
+	}
+	var errors []*Error
+	execution := g.getExecutionString(r)
+	if isCleanStart {
+		errors = g.verifyCleanStart(execution)
+	} else {
+		errors = g.verifyRecovery(execution)
+	}
+	if errors == nil {
+		return true, nil
+	}
+	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
+}
+
+// verifyCleanStart verifies if a specific execution is a valid clean-start execution.
+func (g *GrammarChecker) verifyCleanStart(execution string) []*Error {
+	var errors []*Error
+	lexer := clean_start_lexer.New([]rune(execution))
+	_, errs := clean_start_parser.Parse(lexer)
+	for _, err := range errs {
+		exp := []string{}
+		for _, ex := range err.Expected {
+			exp = append(exp, ex)
+		}
+		expectedTokens := strings.Join(exp, ",")
+		unexpectedToken := err.Token.TypeID()
+		e := &Error{
+			description: fmt.Sprintf("Invalid clean-start execution: parser was expecting one of [%v], got [%v] instead.", expectedTokens, unexpectedToken),
+			height:      err.Line - 1,
+		}
+		errors = append(errors, e)
+	}
+	return errors
+}
+
+// verifyRecovery verifies if a specific execution is a valid recovery execution.
+func (g *GrammarChecker) verifyRecovery(execution string) []*Error {
+	var errors []*Error
+	lexer := recovery_lexer.New([]rune(execution))
+	_, errs := recovery_parser.Parse(lexer)
+	for _, err := range errs {
+		exp := []string{}
+		for _, ex := range err.Expected {
+			exp = append(exp, ex)
+		}
+		expectedTokens := strings.Join(exp, ",")
+		unexpectedToken := err.Token.TypeID()
+		e := &Error{
+			description: fmt.Sprintf("Invalid recovery execution: parser was expecting one of [%v], got [%v] instead.", expectedTokens, unexpectedToken),
+			height:      err.Line - 1,
+		}
+		errors = append(errors, e)
+	}
+	return errors
+}
+
+// addHeightNumbersToTheExecution adds height numbers to the execution. This is used just when printing the execution so we can find the height with error more easily.
+func (g *GrammarChecker) addHeightNumbersToTheExecution(execution string) string {
+	heights := strings.Split(execution, "\n")
+	s := ""
+	// assert that last element of heights == "" (if with error) 
```

Comment:
Please change this comment by an "if" that handles the case where the last element is not ""